### PR TITLE
Update thrust::complex headers with a bug fix

### LINKED ISSUE #2629: Sync the headers in `cupy/core/include/cupy/complex/` with upstream?
Recently the bundled Thrust complex headers have been modified on a per-PR basis for the following reasons:
1. Performance impact: #2562 (fixed alignment)
2. Missing `__host__` qualifiers: #2538
3. Missing support: #2520 

All of these changes are more or less in sync with the latest Thrust codes. I suggest that these headers can be brought up to date with the upstream. The incentive is that one might discover another one-liner change like in #2562 that boosts the performance considerably 🙂

### PR DESCRIPTION
**Pull request by leofang**
Closes #2629. Changes (see the discussion starting https://github.com/cupy/cupy/issues/2629#issuecomment-558448086):
1. Add a `__host__` qualifier to all functions
2. Sync a few function implementations with the upstream counterparts
3. Adjust some style and format
4. Fix a bug: https://github.com/cupy/cupy/pull/2741#discussion_r353360865

### CODE REVIEW

**Comment by emcastillo** [comment]
Regarding  the `__host__` qualifiers.
I understand the value of doing it for diffing files with thrust master. However, since this code is going to be device-only, I am a bit reluctant to it.

**Comment by leofang** [comment]
You may be surprised, but it's needed in supporting external libraries like CUB. (The compilation just wouldn't pass without `__host__` for some reason.) We've managed to make the change only when needed, but it's annoying and would be good to be kept up-to-date.

File: cupy/core/include/cupy/complex/arithmetic.h [hunk 1]
```diff
@@ -155,8 +153,12 @@ __host__ __device__ inline bool operator>(const complex<T>& lhs,
                                           const complex<T>& rhs) {
   if (lhs == rhs) {
       return false;
+  } else if (lhs.real() > rhs.real()) {
+      return true;
+  } else if (lhs.real() == rhs.real()) {
+      return lhs.imag() > rhs.imag();
   } else {
-      return !(lhs < rhs);
+      return false;
```
Comment (leofang) [inline cupy/core/include/cupy/complex/arithmetic.h]: This change fixed a bug:
```python
>>> import cupy as cp
>>> a = cp.asarray([cp.complex(cp.nan, 0.9), cp.complex(0.7, cp.nan)])
>>> a[0] < a[1]  # correct
array(False)
>>> a[0] > a[1]  # wrong, should be False
array(True)
```

File: cupy/core/include/cupy/complex/arithmetic.h [hunk 2]
```diff
@@ -24,272 +24,270 @@ namespace thrust {
...
+  } else if (lhs.real() > rhs.real()) {
+      return true;
+  } else if (lhs.real() == rhs.real()) {
+      return lhs.imag() > rhs.imag();
+  } else {
       return false;
-  } else
-  {
-      return !(lhs < rhs);
```
Comment (leofang) [inline cupy/core/include/cupy/complex/arithmetic.h]: This change fixed a bug:
```python
>>> import cupy as cp
>>> a = cp.asarray([cp.complex(cp.nan, 0.9), cp.complex(0.7, cp.nan)])
>>> a[0] < a[1]  # correct
array(False)
>>> a[0] > a[1]  # wrong, should be False to match NumPy
array(True)
```

**Comment by emcastillo** [comment]
>> You may be surprised, but it's needed in supporting external libraries like CUB

😂awesome. No problem with keeping these then.

**Comment by leofang** [comment]
@emcastillo PTAL.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**